SOLR-14765: Optimize DocList creation by skipping sort for sort-irrelevant cases

magibney opened the pull request:
See: [SOLR-14765](https://issues.apache.org/jira/browse/SOLR-14765)

magibney commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -2040,19 +2098,7 @@ protected void sortDocSet(QueryResult qr, QueryCommand cmd) throws IOException {
       leafCollector.collect(doc - base);
```
@sonatype-lift ignore
This is actually a revert, so unrelated to this PR; in any event I trust this will not manifest in practice as an NPE

madrob commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -1384,16 +1385,24 @@ private void getDocListC(QueryResult qr, QueryCommand cmd) throws IOException {
+        final SortField[] sortFields;
+        // We only need to sort if we're returning results AND sorting by something other than SCORE (sort by
+        // "score" alone is pointless for these constant score queries)
+        needSort = cmd.getLen() > 0 && (sort = cmd.getSort()) != null && ((sortFields = sort.getSort()).length > 1 || sortFields[0].getType() != Type.SCORE);
```
This complex conditional with nested assignments is fairly difficult to read, can we expand to something less terse?

magibney replied:
:+1: ... refactored this (and a couple of other) complex 1-liner conditionals into methods with more transparent names.

magibney commented:
Added a new test suite (2 tests) that verifies the new sort-skipping optimization, and also folds in explicit tests for the useFilterForSortedQuery behavior (as newly updated), which wasn't really directly covered by existing tests.

Pending further feedback, the only thing I _know_ needs to happen is change the "fix version" under which the CHANGES.txt entry is registered.

madrob commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -1299,6 +1302,41 @@ public DocList getDocList(Query query, List<Query> filterList, Sort lsort, int o
 
+  private static boolean isConstantScoreQuery(Query q) {
+    if (q instanceof BoostQuery) {
+      // ConstantScoreQueries are often (always?) wrapped in BoostQuery to assign specific score
```
I suspect this partially goes away after #529

magibney replied:
Something to keep an eye out for; I applied a patch for #529 and this wrapping still seems to happen. I think this is a side-effect of the way "ConstantScoreQuery" gets specified in queries (`(clause)^=n`) where `n` is the constant score, which is applied via wrapping in a BoostQuery. In any event it doesn't look like that changes with #529

madrob replied:
Ok, thanks for checking. We might also want to check for `WrappedQuery`? Not sure.

madrob commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -1299,6 +1302,41 @@ public DocList getDocList(Query query, List<Query> filterList, Sort lsort, int o
+    return sortFields.length > 1 || sortFields[0].getType() != Type.SCORE;
+  }
+
+  private boolean useFilterCacheForDynamicScoreQuery(boolean needSort, QueryCommand cmd) {
```
👍

madrob commented on solr/core/src/test/org/apache/solr/search/TestMainQueryCaching.java:
```diff
@@ -0,0 +1,202 @@
+    String q = "str:d*";
+    String constQ = "("+q+")^=1.0"; // wrapped as a ConstantScoreQuery
+
+    for (int i = 0; i < 6; i++) {
```
An interesting idea I just had was to randomize the order that we run all of these cases. It _shouldn't_ matter because you reload in between but who knows.

The other idea would be to split all of these out into separate test methods, and give them a common assertMetricCounts helper method or something like that.

magibney replied:
:+1: I like the latter idea, I'll give that a shot.

madrob commented on solr/core/src/test/org/apache/solr/core/ExitableDirectoryReaderTest.java:
```diff
@@ -98,7 +102,15 @@ public void testCacheAssumptions() throws Exception {
     // This gets 0 docs back. Use 10000 instead of 1 for timeAllowed and it gets 100 back and the for loop below
     // succeeds.
-    String response = JQ(req("q", "*:*", "fq", fq, "indent", "true", "timeAllowed", "1", "sleep", sleep));
+    // TODO: address crosstalk between test methods; failures here can be triggered by cache consultation
```
Is this still an issue after you added the special MatchAllDocs cases?

magibney replied:
Yes, even more so really, since the special-casing for MatchAllDocs currently just throws it to filterCache. Now that I write this, it occurs to me that if special-casing matchAllDocs maybe we can special-case it to _bypass_ the filterCache entirely. I'll give that a try ...

madrob commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -1405,14 +1446,32 @@ private void getDocListC(QueryResult qr, QueryCommand cmd) throws IOException {
+        out.docList = constantScoreDocList(cmd.getOffset(), cmd.getLen(), out.docSet);
+        if (0 == cmd.getSupersetMaxDoc()) {
+          // this is the only case where `cursorMark && !needSort`
+          qr.setNextCursorMark(cmd.getCursorMark());
```
Is there a test that tries to do `cursorMark=*&rows=0`

magibney replied:
Ahh good catch. Will add one, and for `rows=1` (which should _not_ hit the sort optimization)

magibney commented:
@madrob thanks very much for the review! I believe I've addressed all the concerns you raised. The only one I feel certain might merit more attention is the question of MatchAllDocs special-case, which I've adapted to now not (ever?) hit the filterCache, in 1e6cfeffbabd400b37afe8af8c1bc29d45d8de79. That allowed to revert some of the test workarounds (including a preexisting workaround related to pure-negative filters). Keeping MatchAllDocs off the `filterCache` is I think more generally a Good Thing -- in fact the change addressed an [existing TODO comment](https://github.com/apache/solr/blob/cb2e58f6569704c45387a8d094625ba8e014776b/solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java#L869).

madrob commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -866,8 +876,48 @@ public BitDocSet getLiveDocSet() throws IOException {
+      switch (leafContexts.size()) {
+        case 0:
+          assert numDocs() == 0;
+          docs = new BitDocSet(BitDocSet.empty().getFixedBitSet(), 0);
```
This is confusing about why we need to wrap a `empty()` in a new BitDocSet. Clearer if we use the super type.

```suggestion
          docs = new BitDocSet(DocSet.empty().getFixedBitSet(), 0);
```

madrob commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -866,8 +876,48 @@ public BitDocSet getLiveDocSet() throws IOException {
+          final Bits onlySegLiveDocs = leafContexts.get(0).reader().getLiveDocs();
+          final FixedBitSet fbs;
+          if (onlySegLiveDocs == null) {
+            final int onlySegMaxDoc = maxDoc();
+            fbs = new FixedBitSet(onlySegMaxDoc);
+            fbs.set(0, onlySegMaxDoc);
```
Can you add a comment mentioning the contract of LeafReader.getLiveDocs that if it returns null then we set all the docs live?

madrob commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -866,8 +876,48 @@ public BitDocSet getLiveDocSet() throws IOException {
+            int segOrd = r.maxDoc() - 1;
+            if (segLiveDocs == null) {
+              do {
+                bs.set(segDocBase + segOrd);
```
Can we call the two-argument version of bs.set here? It looks like the Lucene version does some magic to be a bit faster than a straight loop.

magibney replied:
:+1: ha! yeah, not sure why it occurred to me above, but not here.

madrob commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -866,8 +876,48 @@ public BitDocSet getLiveDocSet() throws IOException {
+            } else {
+              do {
+                if (segLiveDocs.get(segOrd)) {
+                  bs.set(segDocBase + segOrd);
```
Same here, two argument version.

magibney replied:
I don't think the 2-arg version works _here_ because the two-arg version works on ranges. In order to call it I think you'd need more complex logic, scanning contiguous ranges and then calling set(int, int) when you encounter boundaries (deleted docs) ... I'm thinking that's not worth the effort here?

madrob commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -866,8 +876,48 @@ public BitDocSet getLiveDocSet() throws IOException {
     // but the addition of setLiveDocs means we needed to add volatile to "liveDocs".
     BitDocSet docs = liveDocs;
     if (docs == null) {
```
Is there a race condition here when multiple match all docs queries come in at the same time? I think that before if there were multiple queries here at the same time then they would all sync on the same result in the filterCache. Hard to tell, this is tricky logic. Probably good to add a test or two?

magibney replied:
You're right that this case kind of undermines (for this special case) the improvements you introduced with SOLR-15555 ... I'll see what I can do about that...

magibney replied:
:+1: this should be addressed now, newly by lazy-initing `liveDocSet` via a single-element special-case CaffeineCache.

madrob commented on solr/core/src/test/org/apache/solr/search/TestMainQueryCaching.java:
```diff
@@ -0,0 +1,227 @@
+/**
+ * Verify caching interactions between main query and filterCache
+ */
+public class TestMainQueryCaching extends SolrTestCaseJ4 {
```
👍

magibney replied:
Yes, checking for `WrappedQuery` wouldn't hurt (would need to respect the cache directives in ExtendedQueryBase then, which would be nice).

madrob commented:
Yea, I'm super excited about this, it's starting to look like a pretty nice boost to what we do!

magibney commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -865,9 +936,26 @@ public BitDocSet getLiveDocSet() throws IOException {
+        }
+      }
+      try {
+        docs = liveDocsFuture.get(); // subsequent callers block here, waiting for initial/only execution to complete
```
resolved this with 779b69b114fd3b7dadb7adf203fd4c7eef5f75ec

dsmiley commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -854,8 +868,58 @@ private DocSet getAndCacheDocSet(Query query) throws IOException {
   private volatile BitDocSet liveDocs;
+  private final FutureTask<BitDocSet> liveDocsFuture = new FutureTask<>(this::computeLiveDocs);
+
+  private BitDocSet computeLiveDocs() {
```
So you wrote all this low level code to avoid a simple MatchAllDocsQuery with a DocSetCollector? See org.apache.solr.search.SolrIndexSearcher.getDocSetNC which calls DocSetUtil.  Even if we choose to keep your low level code, I'd propose it go to a special case inside of DocSetUtil when the query is MatchAllDocsQuery.

magibney replied:
wrt low-level, I realized we needed at least to _check_ for the MatchAllDocsQuery special case in SolrIndexSearcher (in order to siphon that case off before consulting filterCache) -- and having realized that, I basically just ran with [this suggestion/todo comment](https://github.com/apache/solr/commit/1e63b32731bedf108aaeeb5d0a04d671f5663102#diff-99978700f1c69d6a5f6c2190f89c98cfe20c441161db5a183ec002e15cb1be28R857).

I'm ok with losing the special-case low-level code (though it basically makes liveDocSet computation into a glorified array copy -- though to be fair I haven't profiled it). Tangentially: it strikes me that some of this code could be useful in partially restoring cache (e.g. filterCache) entries per-segment sometime down the line ...

That said, although I initially went ahead an moved the code to DocSetUtil as you suggested, I actually ended up moving it back, for reasons explained in the commit message for e368bdd7008894939276f3b94891979cdc4f88b7:
```
    In practice it felt like an awkward (and perhaps somewhat dangerous?)
    fit to have `computeLiveDocs` live in `DocSetUtil`. The potential for
    deadlock was real; and the `liveDocs` concept is really very intimately
    associated with the SolrIndexSearcher per se; putting its computation
    in DocSetUtil might tempt people to call it there, when they should
    really just get liveDocs via `SolrIndexSearcher.getLiveDocSet()`
```
I'm curious to know what you think. I'm find to move it back into DocSetUtil if you think that's better (or for that matter remove it entirely if you prefer).

dsmiley commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -1299,6 +1385,47 @@ public DocList getDocList(Query query, List<Query> filterList, Sort lsort, int o
   public static final int GET_DOCLIST = 0x02; // get the documents actually returned in a response
   public static final int GET_SCORES = 0x01;
 
+  private static boolean isConstantScoreQuery(Query q) {
```
I've seen a need for this elsewhere.  Let's make it handle other cases too (e.g. MatchAllDocsQuery, MatchNoDocsQuery, "Filter" (soon to be DocSetQuery)) and be recursive to the cases below (hey why not).  This could move to QueryUtils rather than having the massive SolrIndexSearcher grow

magibney replied:
:+1: done in dfff5f792a52ff4c8ed87d4495745340c7af10e6. Lmk if this is _too_ exhaustive, or if I've missed anything?

magibney replied:
>I've seen a need for this elsewhere

I looked for other places in the codebase where this could be used, and I didn't immediately find anything. Perhaps I missed something? It does indeed seem general-purpose useful (even if only otherwise leveraged from 3rd-party plugins).

dsmiley replied:
`org.apache.solr.search.QueryUtils#combineQueryAndFilter` wraps a ConstantScoringQuery but it could avoid it.

magibney replied:
FYI, I had to revert the above change; see: c9ccf5c8be98a169742ffc924dda24aa8d3bc4e0

dsmiley replied:
was that uncovered by a test (which)?

magibney replied:
Yes (sorry, I mentioned it in the commit):
```
gradlew :solr:core:test --tests "org.apache.solr.search.TestPseudoReturnFields.testMultiValued"
```

magibney replied:
coincidentally enough I think this may be somewhat related to your comment on [PR #529](https://github.com/apache/solr/pull/529#discussion_r803130966)? Given that this issue has some TODOs related to SOLR-12336 anyway, I'll probably just wait to merge this until SOLR-12336 lands on main and 9_x.

dsmiley replied:
Thanks; I'll look into this after that PR is merged.

dsmiley commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -865,9 +929,31 @@ public BitDocSet getLiveDocSet() throws IOException {
+      } else {
+        // another thread has already called `computeLiveDocs.run()`; this thread will block on
+        // `liveDocsFuture.get()` (below)
+        if (matchAllDocsCacheComputationTracker.getAndIncrement() < 0) {
```
The Future + AtomicLong thing looks overly complicated; maybe I'm not getting why it needs to be so complicated.  Can we have a simple double-check lock with synchronized to ensure only one creation happens?  We've all seen this pattern many times, I'm sure.  We don't even need the liveDocs to be volatile since it's an immutable object, not unlike a String, except for the lazy populated size but that doesn't matter as it's idempotent.

magibney replied:
:+1: ... any reasons I might have had for doing it this way are obviated by following your suggestion below, which just leans on a single-element CaffeineCache to handle the concurrency issues.

dsmiley commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -2281,6 +2455,11 @@ public void initializeMetrics(SolrMetricsContext parentContext, String scope) {
+    parentContext.gauge(skipSortCount::get, true, "skipSortCount", Category.SEARCHER.toString(), scope);
+    parentContext.gauge(matchAllDocsCacheConsultationCount::get, true, "matchAllDocsCacheConsultationCount", Category.SEARCHER.toString(), scope);
+    parentContext.gauge(matchAllDocsCacheHitCount::get, true, "matchAllDocsCacheHitCount", Category.SEARCHER.toString(), scope);
+    parentContext.gauge(matchAllDocsCacheComputationTracker::get, true, "matchAllDocsCacheComputationTracker", Category.SEARCHER.toString(), scope);
```
Is there value in publishing that "Tracker"?  It doesn't even sound like a metric.

If we want these metrics, perhaps we should instead have a one-element Cache, which already publishes metrics and in a standard way across other Caches.  This is just an idea... if it's not easy then nevermind.

magibney replied:
Yes, I went the 1-element cache route and I like it! Way easier and cleaner, I think. I did introduce a couple of additional metrics, including `asyncHits` directly in CaffeineCache, to track the impact of "async" operation (see SOLR-15555). Basically the point of all of these metrics were to "whitebox"-test assumptions about how caching/sort optimizations work. None of them are strictly necessary, but I think they could be useful?

dsmiley reviewed (commented):
On the inter-relationship between DocSetUtil and SolrIndexSearcher, yeah I get what you're saying; there is/was awkward avoidance of an infinite loop to compute the live docs.  I've wondered in the past about perhaps trying to have DocSetUtil have a more formal relationship to SolrIndexSearcher in the way SolrDocumentFetcher does.  SDF once didn't exist but I saw way too much document stuff going on in SIS which is huge.  Perhaps DocSetUtil should be an instance of SolrIndexSearcher tasked with DocSet management (to include live docs), maybe renamed DocSets (plural) or something.  Just an idea... SIS is so long.

magibney commented:
Dropped recursing into BooleanQuery in 9c62f9f5f919ecd1c263f13b16c9e89edd1a84a1.

magibney commented:
re: inter-relationship between DocSetUtil and SolrIndexSearcher.

I see the general issue you're concerned about, but if it's ok I'd like to leave `computeLiveDocs` stuff in SIS for this PR, and if/when work on refactoring stuff out as a separate issue? From my perspective, leaving `computeLiveDocs` in SIS for now really boils down to keeping the method `private`, with the goal of limiting the scopes from which that method is directly accessible -- or really I suppose more trying to ensure that there's exactly _one_ fundamental/canonical way to get liveDocSet, and that is directly via `SIS.getLiveDocSet()`.

magibney commented on solr/core/src/java/org/apache/solr/search/QueryUtils.java:
```diff
@@ -54,23 +54,7 @@ public static boolean isNegative(Query q) {
-   * @param seen used to detect possible loops in nested query input
-   */
-  private static boolean isConstantScoreQuery(Query q, Map<Query, Void> seen) {
+    Map<Query, Void> seen = null; // lazy-init; this will be unnecessary in many cases
```
Oops, was adding to `seen` but never checking it. The purpose should be clarified by 6074c5dc80a1984de74254b88fa7a748c1624735.

It is possible in theory to create a `WrappedQuery` that at some level wraps itself (perhaps indirectly). I'd be more than happy to get rid of this ... do you think it would be safe to do so?

dsmiley replied:
Yes definitely; it's simply wrong to create a circular reference in the first place.  It's very unreasonable to expect every consumer to have to check for this.

magibney replied:
This comment is also relevant:
```
// NOTE: Neither WrappedQuery class nor its inner query are final, so there is a risk of direct loops
// TODO: Only the queries we explicitly check for in this method are relevant wrt detecting loops, and
//  only `WrappedQuery` currently presents a risk in that respect; we may be able to avoid this risk
//  by more tightly restricting the `WrappedQuery` API (e.g., making the get/set methods `final`)?
```

dsmiley replied:
I'd remove these comments as well.  I've never seen this loop checking done before for Queries and I've never seen a Query try to prevent this.

magibney commented:
Thanks for the reviews, @madrob and @dsmiley! Pending further feedback, I hope to fix up the CHANGES.txt entry and commit to main tomorrow. Any thoughts on whether fix version 9.0 or 9.1?

dsmiley commented:
It's internal and not particularly impactful to any plugin writer so I suggest 9.1.

madrob commented:
Interesting stuff going on over in LUCENE-10418 as well that maybe this could compliment in the future?

madrob commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -883,16 +981,30 @@ public Bits getLiveDocsBits() throws IOException {
+   * the DocSet instance supplied), allowing more efficient memory use.
+   * @lucene.internal
+   */
+  public BitDocSet offerLiveDocs(Supplier<DocSet> docSetSupplier, int suppliedSize) {
```
very clear javadoc, this makes sense, thank you!

We don't currently have the convention to do so, but what do you think about annotating this with `com.google.errorprone.annotations.@CheckReturnValue` so that it would cause compilation error if the return value was ignored?

madrob commented on solr/core/src/test/org/apache/solr/search/TestMainQueryCaching.java:
```diff
@@ -0,0 +1,313 @@
+    createIndex();
+  }
+
+  @AfterClass
+  public static void afterClass() throws Exception {
+    if (RESTORE_UFFSQ_PROP == null) {
+      System.clearProperty(TEST_UFFSQ_PROPNAME);
+    } else {
+      System.setProperty(TEST_UFFSQ_PROPNAME, RESTORE_UFFSQ_PROP);
+    }
+  }
```
The test framework takes care of this for us already, but it's fine to leave it here.

madrob commented on solr/core/src/test/org/apache/solr/search/TestMainQueryCaching.java:
```diff
@@ -0,0 +1,313 @@
+
+  @BeforeClass
+  public static void beforeClass() throws Exception {
+    // TODO: figure out why the line below (accepting this property as overridden on test invocation) isn't working
+    //  as expected.
```
What's the issue here? I'm not entirely sure what I'm looking for.

magibney replied:
The goal was to allow explicitly specifying `-Psolr.test.useFilterForSortedQuery=true|false` on the command line, but it looks like the system properties are filtered by gradle maybe (?) ... in any event when I tried to set this property explicitly it didn't get picked up in the actual test. If there's a quick answer, I'd be curious to know; but either way I agree with your implicit suggestion that leaving this as a TODO comment doesn't make sense ...

magibney replied:
Removed this with d87801882fae79b391080cee735dfc615e585fbf; it's really orthogonal to this PR

madrob commented on solr/core/src/test/org/apache/solr/search/TestMainQueryCaching.java:
```diff
@@ -0,0 +1,313 @@
+  public void testMatchAllDocsFlScore() throws Exception {
+    // explicitly requesting scores should unconditionally disable all cache consultation and sort optimization
+    String response = JQ(req("q", MATCH_ALL_DOCS_QUERY, "indent", "true", "rows", "0", "fl", "id,score", "sort", (random().nextBoolean() ? "id asc" : "score desc")));
+    // NOTE: pretend we're not MatchAllDocs ...
```
What does this mean? Did you want to do an exists query that happens to match all docs, like `str:*` (or `id:*`)

magibney replied:
Yeah this is confusing. No we actually want `*:*` here I think, because we're testing that even when `*:*` is specified, if the user _explicitly_ requests that `score` be returned, that disables the optimizations we have otherwise. This is an edge-case-of-an-edge-case, and probably an antipattern (to explicitly request `score` for `*:*`). But one practical consequence of this for testing it's useful to be able to bypass the optimization.

magibney replied:
added a clarifying comment in b494419ca1eff547a73ecf81a608cb6abaa789a3

madrob commented on solr/core/src/test/org/apache/solr/search/TestMainQueryCaching.java:
```diff
@@ -0,0 +1,313 @@
+    }
+    try {
+      for (Future<?> f : followup) {
+        f.get(); // to access exceptions/errors
```
`assertNull` here

magibney replied:
I'm curious why? If the issue is _swallowing_ errors/exceptions, these would not be returned by `.get()` anyway, but will instead be thrown wrapped in `ExecutionException`, iiuc.

madrob commented on solr/core/src/test/org/apache/solr/search/TestMainQueryCaching.java:
```diff
@@ -0,0 +1,313 @@
+    }
+    final SolrCore core = h.getCore();
+    Map<String, Object> liveDocsCacheMetrics = coreToLiveDocsCacheMetrics(core);
+    long inserts = (long) liveDocsCacheMetrics.get("inserts"); // the one and only liveDocs computation
+    long hits = (long) liveDocsCacheMetrics.get("hits"); // hits during the initial phase
+    long asyncHits = (long) liveDocsCacheMetrics.get("asyncHits");
```
use the constants in SolrCache instead of strings here?

dsmiley replied:
FWIW I prefer readability in tests (String literals are plainly obvious).  And Yonik once pointed out that avoiding such constants in tests helps ensure we don't change something without seeing the impact (e.g. a client may depend on it).

madrob replied:
Yea, I can go either way. My thought was that if the constant ever moves, or the metric path moves, we'd want tests to fail compilation until they are updated. But this makes sense too. I'm not committed to either way.

magibney replied:
Interesting; I'm really ambivalent :-| ... I can see good arguments both ways. I'm inclined to let inertia win and leave it as-is?

madrob commented on solr/core/src/test/org/apache/solr/search/TestMainQueryCaching.java:
```diff
@@ -0,0 +1,313 @@
+    // It seems that the assertion below more frequently succeeds when this test is run in isolation; e.g.:
+    // `gradlew :solr:core:test --tests "org.apache.solr.search.TestMainQueryCaching.testConcurrentMatchAllDocsInitialization"`
+
+    //assertTrue("expected asyncHits > 0; found asyncHits=" + asyncHits, asyncHits > 0);
```
We could introduce delay at some point to cause threads to stall via TestInjection, but I'm fine without it.

magibney replied:
That'd be really cool. I actually kind of knocked myself out a little trying to do something like that but didn't want to reinvent the wheel or end up with something that needlessly slows down the test suite. If you know of a good (dare I say "right"?) way to do this I'm definitely interested.

magibney commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -119,10 +128,18 @@
+  /**
+   * Special-case cache to handle the lazy-init of {@link #liveDocs}.
+   */
+  private final SolrCache<MatchAllDocsQuery,BitDocSet> liveDocsCache;
```
Yeah; it started out as something ostensibly "simpler" like that, but once you layer in the metrics (mainly to support tests) it becomes less simple. I wonder how succinct it could be, maybe worth a shot.

I'll mull this; if it's something you feel (somewhat) strongly about I'll take a crack at it.

madrob replied:
I don't feel strongly about it, and would be satisfied with an additional line in the comment saying why it is what it is.

magibney commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -854,20 +880,92 @@ private DocSet getAndCacheDocSet(Query query) throws IOException {
+    }
+  }
+
+  private static void copyTo(Bits segLiveDocs, int sourceMaxDoc, FixedBitSet bs, int segDocBase) {
```
Absolutely, thanks for raising that point. I'll make sure to add something before committing this.

magibney replied:
Ok, added a unit test for this (to TestDocSet.java) in 8defddff3e18906ac03f21ea85c91405c516397e.

magibney commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -854,20 +880,92 @@ private DocSet getAndCacheDocSet(Query query) throws IOException {
+
+  private static final BitDocSet EMPTY = new BitDocSet(new FixedBitSet(0), 0);
+
+  private BitDocSet computeLiveDocs(Query q) {
+    assert q == MATCH_ALL_DOCS_QUERY;
```
I guess the thought was that this method looks tempting to call ... "computeLiveDocs? Sure I wanna do that!". It's really supposed to be very locked-down in terms of the contexts from which it's called (and even then not called directly -- it's called via CaffeinCache.computeIfAbsent).

Reverse-engineering my subconscious thought process here: normally `private` access modifier alone would take care of that, but in a class as sprawling as SIS the power/semantics of `private` risk being diluted. Could change the method signature to MatchAllDocsQuery, yes; but paradoxically that might make it seem _more_ like it can be called by anyone with a MatchAllDocsQuery.

If this logic makes sense I'll add a comment explaining this in a more succinct way. I don't mind removing the assertion, if you prefer.

magibney replied:
obviated by simplifications related to 14f7ac5434626e5d3a068b8c2489b0d604171daa

magibney commented:
Thanks for the feedback, @madrob. I think everything should be addressed now except for the [unit test](#discussion_r806048246) for `copyTo(...)`, which will take a bit more time but should be fairly cut-and-dry as far as what's required.

magibney commented:
As a result of 14f7ac5434626e5d3a068b8c2489b0d604171daa, we could revert the addition of the `asyncHits` metric to `CaffeineCache.java`. I think it's a potentially useful metric? But Mike, David, unless you think it should stay I'm inclined to remove it from this PR (and tbh I probably won't pursue adding it without some renewed motivation to do so).

magibney commented:
I'm hoping to commit this this week.

I did want first to specifically call out that _after_ @dsmiley's PR approval, there were changes to the way the liveDocs DocSet is cached: basically, no longer uses a 1-element CaffeineCache and instead handles synchronization manually within SIS, adding a few metrics mainly to support test assertions. I'm comfortable with this approach and pretty sure the concurrency stuff is correct and idiomatic, but in case anyone wants to take another look before I commit, the substance of the new approach is mostly at [SIS.populateLiveDocs](https://github.com/apache/solr/blob/91222454725122a76399348eb3143e373d2b222c/solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java#L1007-L1035).

Thanks again!

dsmiley commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -935,22 +956,96 @@ private DocSet getAndCacheDocSet(Query query) throws IOException {
+      }
+    }
+    final BitDocSet docs;
+    if (computeInline) {
+      docs = liveDocsSupplier.get();
+      liveDocsCacheInstance.complete(docs);
+      liveDocs = docs;
+      liveDocsInsertsCount.increment();
+    } else {
```
Why not do this in the synchronized block above?  It's natural that it would go there because we're initializing it -- a standard pattern to use synchronized in this way.  Yes it would cause callers to block (contention) but it's fair/expected.  They will any way, waiting on the CompletableFuture.

magibney replied:
Well I guess the only downside of running the computation within the synchronized block is that you lose the metrics distinction between `liveDocsHitCount` and `liveDocsAsyncHitCount`. That's probably mainly a curiosity now ("during the initialization of liveDocs, approximately how many redundant livedocs computations were avoided?"); the initial hope was to use that metrics distinction in tests to verify that requests which _would_ have resulted in redundant livedocs computations were in fact waiting on a single computation. But I kind of dropped that -- see [here](https://github.com/apache/solr/blob/91222454725122a76399348eb3143e373d2b222c/solr/core/src/test/org/apache/solr/search/TestMainQueryCaching.java#L445-L460) and [here](#discussion_r806102113).

If deciding to move computation inside the synchronized block, I think it'd make sense to remove the metrics distinction, since iiuc it would no longer be measuring anything meaningful.

madrob replied:
If we do everything in the synchronized block, then there's no reason for the CompletableFuture at all anymore, right? I feel like at that point we're losing concurrency gains, but would have to think about it more to be sure.

magibney replied:
I think you're right. And I think you don't even lose on concurrency gains:

In both cases execution blocks until the result is available. It doesn't matter whether you're blocking for a synchronized lock or on getting results from the Future. The only way it matters is if you're attempting to make a distinction between "I'm waiting because a computation is in progress" and "I had to get a lock on the object that will ensure computation has completed". So if we lose the "asyncHits" metric (I'm fine with that), then things can be much simpler.

Tried this out with b3b38a1b2e280febd6a4a86464fad36bea4c56b8 ... seems good to me?

dsmiley commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -118,10 +122,22 @@
 
+  /** Special-case cache to handle the lazy-init of {@link #liveDocs}. */
+  @SuppressWarnings({"unchecked", "rawtypes"})
+  private final CompletableFuture<BitDocSet>[] liveDocsCache = new CompletableFuture[1];
```
I would declare it to be simply Future and not specifically CompletableFuture as it's an implementation detail that can be local to the only method that uses it.

BTW is any of that SuppressWarnings actually appropriate here?  I don't see either occuring.  There was a big effort to scrub the codebase of needless suppressions and also declaring things better so that we needn't have a warning in the first place.

magibney replied:
Both SuppressWarnings are necessary with the "internal-one-element-array-for-synchromization" approach, because of the inability to do parameterized array creation. Is there another alternative that I'm missing there?

I'd be ok with Future if you prefer; the only benefit I can see to CompletableFuture on the instance variable (which is private, so arguably is itself an implementation detail, albeit in a very large class :slightly_smiling_face:) is that it provides the `join()` method to get the computation results without forcing us to explicitly handle checked exceptions from the `get()` method.

magibney replied:
>Both SuppressWarnings are necessary

to be more specific: gradle won't compile the project unless I add these two SuppressWarnings.

madrob replied:
You can do with only 1 suppression if you use `Array.newInstance` - there's an example in ZkController

magibney replied:
This thread should also be addressed by b3b38a1b2e280febd6a4a86464fad36bea4c56b8

dsmiley commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -968,19 +1050,23 @@ public Bits getLiveDocsBits() throws IOException {
     // a few places currently expect BitDocSet
-    assert docs.size() == numDocs();
-    this.liveDocs = makeBitDocSet(docs);
+    ret = populateLiveDocs(() -> makeBitDocSet(docSetSupplier.get()));
```
pointless; I'd do this as one line -- just return it

dsmiley commented on solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:
```diff
@@ -935,22 +956,83 @@ private DocSet getAndCacheDocSet(Query query) throws IOException {
+  }
+
+  private BitDocSet populateLiveDocs(Supplier<BitDocSet> liveDocsSupplier) {
+    final BitDocSet inlineDocs;
```
As I suggested in my comment above, this can be simplified with merely an Object used to synchronize.  No need for a "inlineDocs" local variable.  No need for an array or anything similar-is.  Simplicity!

magibney commented:
Thanks, @dsmiley! I think ef4e35e065be8385a3ca7219df471d68bb567c51 addresses all these concerns.

dsmiley reviewed (approved):
Thanks Michael; I appreciate you satisfying my requests for ruthless simplification :-)
